### ISSUE TITLE
Regression (osism-v5.x): Booting from vda (virtio) breaks

### ISSUE THREAD

**Issue by garloff** [osism/issues#509]
On wavestack, booting from a volume that previously succeeded no longer works.
Booting the "Ubuntu Minimal 22.04" image to a flavor with local boot disk does no longer work either.
The noVNC console shows that it hangs in the BIOS, trying to load the bootloader from disk.
After spending half a day, the issue could be tracked to the image having set `hw_disk_bus=virtio` (vda).
Changing the image-property to `hw_disk_bus=scsi` (sda) makes it work again.
This is likely a regression in OSISM-5.x -- the VM in question had been running for a while before.

Sidenote: The [SCS image metadata standard](https://github.com/SovereignCloudStack/standards/blob/main/Drafts/Image-Properties-Spec.md) does recommend SCSI, as it tends to perform slightly better and supports DISCARD (fstrim). vda is common enough that we can probably not ignore this, though.

**Comment by berendt** [comment]
At least in the image manager we have the following definition (which should be fine):

https://github.com/osism/openstack-image-manager/blob/main/etc/images/ubuntu.yml#L247-L249

**Comment by berendt** [comment]
The following entry is included in the quay.io/osism/osism:0.20230407.0 image (this one is part of the 5.1.0 release). Looks good there as well.

```
  - name: Ubuntu 22.04 Minimal
    enable: true
    shortname: ubuntu-22.04-minimal
    format: qcow2
    login: ubuntu
    min_disk: 8
    min_ram: 512
    status: active
    visibility: public
    multi: true
    meta:
      architecture: x86_64
      hw_disk_bus: scsi
      hw_rng_model: virtio
      hw_scsi_model: virtio-scsi
      hw_watchdog_action: reset
      os_distro: ubuntu
      os_version: '22.04'
      replace_frequency: quarterly
      uuid_validity: last-3
      provided_until: none
    tags: []
    latest_checksum_url: https://cloud-images.ubuntu.com/minimal/releases/jammy/release/SHA256SUMS
    latest_url: https://cloud-images.ubuntu.com/minimal/releases/jammy/release/ubuntu-22.04-minimal-cloudimg-amd64.img
    versions:
      - version: '20230328'
        url: https://minio.services.osism.tech/openstack-image-manager/ubuntu-22.04-minimal/20230328-ubuntu-22.04-minimal.qcow2
        checksum: sha256:04dd2f3ed2ed76f4a3563cacada000bfe0409b959fd7e4dd5e72c80953c186c8
        build_date: '2023-03-28'
```

**Comment by garloff** [comment]
Wavestack seems to have changed hw_disk_bus, so they can attach more disks.
I don't think that's a good idea (and we should rather address it if there are limitations with the no of SCSI disks that can be attached.)
That said: Booting with hw_disk_bus=virtio works on most OpenStack clouds and used to work for SCS as well, AFAIK. So breaking it is not good. We can have a discussion whether we want to deprecate it...

**Comment by chschilling** [comment]
I am talking to my team, why we use virtio in that place. Could be that it is a leftover from pre image manager.
But I agree with @garloff 
virtio as default on many environments shouldn't be a game breaker.

**Comment by chschilling** [comment]
Actually, I can't confirm that it is a virtio or scsi issue.
I tested @garloff ubuntu minimal with scsi and ours with virtio.
Both refused to start and are stuck in BIOS.

Other images are booting without any issues, even with virtio in metadata

**Comment by linwalth** [comment]
> Wavestack seems to have changed hw_disk_bus, so they can attach more disks. I don't think that's a good idea (and we should rather address it if there are limitations with the no of SCSI disks that can be attached.) That said: Booting with hw_disk_bus=virtio works on most OpenStack clouds and used to work for SCS as well, AFAIK. So breaking it is not good. We can have a discussion whether we want to deprecate it...

virtio is not needed anymore.

https://docs.openstack.org/nova/latest/configuration/config.html#compute.max_disk_devices_to_attach

**Comment by garloff** [comment]
> Both refused to start and are stuck in BIOS.

I can't confirm this.
* The public image from wavestack ("Ubuntu Minimal 22.04") with `hw_disk_bus=virtio` does not boot
* My privately registered image (from the same upstream .img file!) "Ubuntu 22.04 Minimal" with `hw_disk_bus=scsi` boots (tested on SCS-1L-1-5) without problems

**Comment by garloff** [comment]
>     * The public image from wavestack ("Ubuntu Minimal 22.04") with `hw_disk_bus=virtio` does not boot
> 
>     * My privately registered image (from the same upstream .img file!) "Ubuntu 22.04 Minimal" with `hw_disk_bus=scsi` boots (tested on SCS-1L-1-5) without problems

OK, so maybe we need to discard all of this.
@chschilling looked into this more closely.
* Ubuntu 22.04 Minimal does not display anything so the noVNC console looks like no booting did happen
  - It does create a serial console log, which I could not get to yesterday, but now can. (I always do console log show before falling back to console url show.)
  - `console=tty console=ttyS0` boot parameters should result in output to both VGA console and serial console, but for some reason the kernel may not have the right VGA driver / EFIFB ... ?
* So the conclusion may have been mislead by not being able to access the VM via network and not seeing anything intelligible on the noVNC console.
* Retesting today, with the ability to get the serial console log and to have working floating IPs, I see both the `hw_bus_type=virtio` (vda) boot "Ubuntu Minimal 22.04" succeeding (on SCS-1V-4-10) as well as the `hw_bus_type=scsi` (sda) boot "Ubuntu 22.04 Minimal" (on SCS-1L-1-5).
* We don't know whether some boots from vda might have been successful without us being able to detect it yesterday or whether some fix from last night or today has addressed the ability to boot from vda again.

**Comment by artificial-intelligence** [comment]
you should still investigate why scsi is used, apart from using many devices, which is a fine usecase.

just to reiterate:

- virtio-blk supports discard just fine, no need for virtio-scsi
- the claimed performance benefit for virtio-scsi seems to be wrong, see my posting here: https://github.com/osism/openstack-image-manager/issues/564#issuecomment-1542145987

**Comment by garloff** [comment]
@artificial-intelligence :
* I have tested that on Ubuntu 22.04 (5.15 kernel), fstrim appears to work against vda
* I remember this not being the case in the not too old past

I still vote in favor of scsi:
* Device names with sda would be consistent with disks on bare metal nodes
* fstrim/DISCARD working on old kernels
* It is our current recommendation and change always has the potential to cause headaches
* I don't see any downside ... in the end the virtio-scsi driver also uses pretty much the same way of transferring blocks back and forth as the virtio-blk driver does, using the nicely optimized virtio interface

**Comment by artificial-intelligence** [comment]
well, it's of course up to you, the [kernel patch is from 2017](https://patchwork.kernel.org/project/qemu-devel/patch/1490690365-21109-1-git-send-email-changpeng.liu@intel.com/), but I don't know when it got merged (I researched this all two years ago, but don't have access to my old docs, unfortunately).

I had this stuff tested on ubuntu 18.04 with the HWE Kernel (both on compute nodes and afaik in the VM as well).

to sum it up:

- virtio-blk is upstream default
- it's faster in most cases
- you really should not rely for anything important on unstable device names like `/dev/vda` or `/dev/sda` anyway, that device might get another name after the next boot, especially if you attach multiple devices, I have personally seen servers fail to boot which relied on these unstable paths, use Disk Labels or /dev/disk-by-uuid/ path.

But I see the point of not changing defaults and to be able to attach more devices, that's a usecase I have seen also many times in the past.

I'm just in favor of making informed decisions clearly mentioning all up- and downsides regardless the topic, just in case someone complains later :wink: 

So I see your point clearly, imho it should just be noted/documented what are the up- and downsides of each decision, so we can come back later in the future and complain about our poor decision making. I'm only half joking here. :wink:
